### PR TITLE
Add new task verify-enterprise-contract

### PR DESCRIPTION
**Pull request by lcarva**
Notable differences from the existing enterprise-contract.yaml task:

* The only data being validated is the image attestation. No data
  data is fetched from the cluster.
* No support for inline public key. The parameter PUBLIC_KEY is used
  as is. I didn't want to wrap cosign in a bash script just to support
  this. Realistically, the key is going to be stored in the cluster
  anyways, just use that. Do note that the Secret for the key does not
  have to be the same one used by Chains. In fact, users of this
  task are expected to create their own Secret containing just the
  cosign.pub key in their own namespaces.
* The whole enterprise contract is verified by this task. This is the
  same approach used by the existing enterprise-contract.yaml task,
  but differs from the cosign-verify*.yaml tasks.

Signed-off-by: Luiz Carvalho <lucarval@redhat.com>

### CODE REVIEW

File: appstudio-utils/util-scripts/lib/fetch.sh
Comment (simonbaird) [inline appstudio-utils/util-scripts/lib/fetch.sh:27]: Just FYI, I moved this in #52 which isn't merged yet.

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (simonbaird) [inline appstudio-utils/util-scripts/verify-attestation-with-policy.sh:32]: Probably won't need this after https://github.com/hacbs-contract/ec-policies/pull/11 is merged, which might happen in a minute or two...

File: appstudio-utils/util-scripts/lib/fetch.sh
Comment (lcarva) [reply appstudio-utils/util-scripts/lib/fetch.sh:27]: Awesome! Just approved that PR.

File: tasks/verify-enterprise-contract.yaml
Comment (simonbaird) [inline tasks/verify-enterprise-contract.yaml:71]: Oh that's really clean. Good idea to not have messy bash here in the task definition.

File: tasks/verify-enterprise-contract.yaml
Comment (simonbaird) [inline tasks/verify-enterprise-contract.yaml:66]: What are the square brackets for here?

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (simonbaird) [reply appstudio-utils/util-scripts/verify-attestation-with-policy.sh:32]: Though.. the rego rules in there do make no incorrect assumptions about the location of attestations in the $EC_WORK/data directory.

File: tasks/verify-enterprise-contract.yaml
Comment (simonbaird) [reply tasks/verify-enterprise-contract.yaml:71]: Do we use /shared/image-signature.json at any point? It's invisible to opa/rego here, right?

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (simonbaird) [inline appstudio-utils/util-scripts/verify-attestation-with-policy.sh:38]: So the predicate *will* be visible to opa/rego under `data.predicate`, right?
This means we can't write a policy to validate the attestation _type or predicateType or the subject.
Not suggesting it's a problem, but just noting it. Is that intentional design choice? My instinct might be to throw in the full attestation in case we do want it visible in rego, but I'm not confident that's wise or correct.

(One of the (placeholder/demo) rego rules currently does look at the `_type` value at the top level of the attestation. We can easily ditch it though.)

File: tasks/verify-enterprise-contract.yaml
Comment (lcarva) [reply tasks/verify-enterprise-contract.yaml:71]: Yeah, the file is not used anywhere. I should probably just remove the `--output-file` parameter to avoid confusion.

File: tasks/verify-enterprise-contract.yaml
Comment (lcarva) [reply tasks/verify-enterprise-contract.yaml:66]: `command` expects a list, thus the square brackets. The [docs](https://tekton.dev/docs/pipelines/tasks/) do not explain why `command` is a list, but every example does follow the pattern of a single item in the `command` list, and other parts going into the `args` list. I tried using a string instead, and the Tekton rejects it.

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (lcarva) [reply appstudio-utils/util-scripts/verify-attestation-with-policy.sh:32]: True. I can play around with that repo as a way to get this PR further along.

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (lcarva) [reply appstudio-utils/util-scripts/verify-attestation-with-policy.sh:38]: `_type` and `predicateType` are there. They are part of what gets decoded from the payload.

I think looking at these two fields from rego will be important for us.

File: tasks/verify-enterprise-contract.yaml
Comment (lcarva) [reply tasks/verify-enterprise-contract.yaml:70]: Remove the parameter.

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (lcarva) [reply appstudio-utils/util-scripts/verify-attestation-with-policy.sh]: With that PR, this is no longer needed. Removed it. (I'll still follow up on adjusting ec-policies further)

File: appstudio-utils/util-scripts/lib/fetch.sh
Comment (simonbaird) [reply appstudio-utils/util-scripts/lib/fetch.sh:27]: Merged. You can rebase when ready.

File: tasks/verify-enterprise-contract.yaml
Comment (simonbaird) [reply tasks/verify-enterprise-contract.yaml:66]: Ack, thanks.

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (simonbaird) [reply appstudio-utils/util-scripts/verify-attestation-with-policy.sh:29]: As per slack discussion, it makes sense to me now, thanks.

There's some potential for confusion around the terminogy, i.e. payload vs attestation vs predicate vs envelope and etc, but we'll figure it out.

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (simonbaird) [inline appstudio-utils/util-scripts/verify-attestation-with-policy.sh:21]: typo: just noticed it's missing an 'a'

    attestation

File: tasks/verify-enterprise-contract.yaml
Comment (simonbaird) [inline tasks/verify-enterprise-contract.yaml:54]: Not sure either, other than it comes from https://github.com/redhat-appstudio/build-definitions/blob/main/appstudio-utils/Dockerfile .

File: tasks/verify-enterprise-contract.yaml
Comment (simonbaird) [inline tasks/verify-enterprise-contract.yaml:30]: I'd consider default to "1" here, or even just leave it hard coded down below and remove the param.

**Comment by simonbaird** [comment]
Does it need a version of hack/start-ec-task.sh ?

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (simonbaird) [inline appstudio-utils/util-scripts/verify-attestation-with-policy.sh:35]: Same thought as in the ec-policies PR. Would there be multiple of them? Or at least, would the content of that file be an array? If not then maybe "attestation" is a better filename here.

File: tasks/verify-enterprise-contract.yaml
Comment (lcarva) [reply tasks/verify-enterprise-contract.yaml:32]: It is now, thanks!

File: tasks/verify-enterprise-contract.yaml
Comment (lcarva) [reply tasks/verify-enterprise-contract.yaml]: I figure it out. The `build-and-push.sh` script will use this special value to dynamically update the value to a reference of the image built from the Dockerfile you mentioned.

File: tasks/verify-enterprise-contract.yaml
Comment (lcarva) [reply tasks/verify-enterprise-contract.yaml]: Changed the default to 1... just in case.

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (lcarva) [reply appstudio-utils/util-scripts/verify-attestation-with-policy.sh]: This is addressed now. `data/attestations/data.yaml` is now always a list of attestations.

**Comment by lcarva** [comment]
> Does it need a version of hack/start-ec-task.sh ?

@simonbaird, yes, I think so. Added.

File: hack/build-and-push.sh
Comment (lcarva) [inline hack/build-and-push.sh:53]: cc @Michkov

Comment (simonbaird) [reply hack/build-and-push.sh:53]: I had to squint pretty hard, but I think I get it now.

File: tasks/verify-enterprise-contract.yaml
Comment (simonbaird) [inline tasks/verify-enterprise-contract.yaml:87]: See also https://issues.redhat.com/browse/HACBS-316 .

File: hack/start-verify-ec-task.sh
Comment (simonbaird) [inline hack/start-verify-ec-task.sh:23]: Seems like a previous version of the script name here.

File: tasks/verify-enterprise-contract.yaml
Comment (lcarva) [reply tasks/verify-enterprise-contract.yaml]: Awesome. Added a link to it for the future.

File: hack/build-and-push.sh
Comment (simonbaird) [inline hack/build-and-push.sh:1]: Intentional?

File: hack/build-and-push.sh
Comment (lcarva) [reply hack/build-and-push.sh:53]: Given the amount of time I spent on this, it's probably one of the most expensive lines I ever wrote.

File: hack/start-verify-ec-task.sh
Comment (lcarva) [reply hack/start-verify-ec-task.sh]: Thanks for catching it! Fixed.

File: hack/build-and-push.sh
Comment (lcarva) [reply hack/build-and-push.sh]: Negative. Fixed it. Thanks!

File: tasks/verify-enterprise-contract.yaml
Comment (simonbaird) [inline tasks/verify-enterprise-contract.yaml:84]: Aah, so I guess there's no way for cosign to learn the rekor host from what in finds in oci.

File: hack/start-verify-ec-task.sh
Comment (simonbaird) [inline hack/start-verify-ec-task.sh:16]: Do you want to throw in an example oc command here for the kids in the back row?

Comment (lcarva) [reply hack/start-verify-ec-task.sh]: Good idea. Added a two-liner.

File: tasks/verify-enterprise-contract.yaml
Comment (lcarva) [reply tasks/verify-enterprise-contract.yaml:84]: Not that I know of.

File: hack/start-verify-ec-task.sh
Comment (simonbaird) [inline hack/start-verify-ec-task.sh:18]: Awesome.

**Comment by simonbaird** [comment]
Here's the new task running (and failing) in my cluster against an image I built a while ago:
http://pastebin.test.redhat.com/1046547

Here's it is passing with a modified policy:
http://pastebin.test.redhat.com/1046548

I think this could be ready to merge.

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (simonbaird) [inline appstudio-utils/util-scripts/verify-attestation-with-policy.sh:41]: Sometimes you can lose the formatting if you use echo without quoting. Maybe we need this:

    echo -n "${attestations}" > $(json-data-file attestations)

Comment (lcarva) [reply appstudio-utils/util-scripts/verify-attestation-with-policy.sh]: Good idea. Done.

**Review by simonbaird — APPROVED** [review]

**Comment by simonbaird** [comment]
It's good, I think merge when ready, though you could consider seeking one more opinion if you like.

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (robnester-rh) [inline appstudio-utils/util-scripts/verify-attestation-with-policy.sh:7]: Lines 5 and 7 should be using the same nomenclature. Either `attestation file` in both, or `in-toto file` in both.

**Comment by lcarva** [comment]
> It's good, I think merge when ready, though you could consider seeking one more opinion if you like.

@simonbaird, I don't have merge permission.

**Review by robnester-rh — APPROVED** [review]
Other than a note regarding parameter naming, good work.

File: appstudio-utils/util-scripts/verify-attestation-with-policy.sh
Comment (simonbaird) [reply appstudio-utils/util-scripts/verify-attestation-with-policy.sh:7]: Dang it. Have to fix it later now..

Comment (lcarva) [reply appstudio-utils/util-scripts/verify-attestation-with-policy.sh:7]: Fixed in https://github.com/redhat-appstudio/build-definitions/pull/62

File: hack/start-verify-ec-task.sh
Comment (zregvart) [inline hack/start-verify-ec-task.sh:61]: I find it better to assign a random name, i.e. `name: verify-enterprise-contract-$(openssl rand --hex 10)` and be sure not to trigger a race condition below

File: hack/start-verify-ec-task.sh
Comment (zregvart) [inline hack/start-verify-ec-task.sh:83]: Using `--last` is problematic, we don't know the number of tasks that will run concurrently so we could be tailing the logs of the wrong task here. We should generate a name for the task, like above, store it in a variable and use it here.

File: tasks/verify-enterprise-contract.yaml
Comment (zregvart) [inline tasks/verify-enterprise-contract.yaml:25]: Not allowing an inline public key can lead to issues when the signing key is rolled over as we assume that the signature was created with the same key pair that is stored in both secrets: `signing-secrets` for chains controller and the `...-cosign-public-key` secret containing only the public key. There is a window where the signature is created using previous key pair and both secrets have been updated to the newest key. Running and retrying the task can become problematic as the time between signing the image/attestation and verifying the signature increases.

File: tasks/verify-enterprise-contract.yaml
Comment (zregvart) [inline tasks/verify-enterprise-contract.yaml:70]: Should be [`SSL_CERT_DIR`](https://github.com/golang/go/blob/96c8cc7fea94dca8c9e23d9653157e960f2ff472/src/crypto/x509/root_unix.go#L25)

**Review by zregvart — COMMENTED** [review]
Come comments that might get addressed in future PRs, here added after merge to best link up with the code.

File: hack/start-verify-ec-task.sh
Comment (lcarva) [reply hack/start-verify-ec-task.sh:83]: This script is under `hack`, it is meant to be used to assist us in developing and testing our task. It's not meant for production workloads.

I do see your point though. Since this script starts the task, it could very well be more deterministic on which task it follows.

File: tasks/verify-enterprise-contract.yaml
Comment (lcarva) [reply tasks/verify-enterprise-contract.yaml:70]: Cool. I found `SSL_CERT_DIR` in the docs as well but was confused because the `cosign-verify*` tasks use `SSL_CERTS_DIR`.

File: hack/start-verify-ec-task.sh
Comment (lcarva) [reply hack/start-verify-ec-task.sh:61]: +1 https://github.com/redhat-appstudio/build-definitions/pull/69

File: hack/start-verify-ec-task.sh
Comment (lcarva) [reply hack/start-verify-ec-task.sh:83]: Fixed in https://github.com/redhat-appstudio/build-definitions/pull/69

File: tasks/verify-enterprise-contract.yaml
Comment (zregvart) [reply tasks/verify-enterprise-contract.yaml:70]: Good catch! That's a typo there, a [pr to fix it](https://github.com/redhat-appstudio/build-definitions/pull/70).

**Pull request merged**
